src/adaptix/_internal/morphing/constant_length_tuple_provider.py

```diff
@@ -0,0 +1,326 @@
+        norm = try_normalize_type(get_type_from_request(request))
+        if len(norm.args) > 1 and norm.args[1] == Ellipsis:
+            raise CannotProvide
+        if HAS_UNPACK and any(arg.origin == typing.Unpack for arg in norm.args if arg != Ellipsis):
+            raise CannotProvide(
+                "typing.Unpack isn't supported yet",
+                is_terminal=True,
+                is_demonstrative=True
+            )
+
+        loaders = mediator.mandatory_provide_by_iterable(
+            [
+                LoaderRequest(
+                    loc_map=LocMap(
+                        TypeHintLoc(type=tp.source)
+                    )
+                )
+                for tp in norm.args
+            ]
+        )
+        strict_coercion = mediator.mandatory_provide(StrictCoercionRequest(loc_map=request.loc_map))
+        debug_trail = mediator.mandatory_provide(DebugTrailRequest(loc_map=request.loc_map))
+        return self._make_loader(tuple(loaders), strict_coercion, debug_trail)
+
+    def _make_loader(self, loaders: Collection[Loader], strict_coercion: bool, debug_trail: DebugTrail):
+        if debug_trail == DebugTrail.DISABLE:
+            if strict_coercion:
+                return self._get_dt_disable_sc_loader(loaders)
+            return self._get_dt_disable_non_sc_loader(loaders)
+        if debug_trail == DebugTrail.FIRST:
+            tuple_mapper = self._create_dt_first_loader(loaders)
+        elif debug_trail == DebugTrail.ALL:
+            tuple_mapper = self._create_dt_all_loader(loaders)
+        else:
+            raise ValueError
+        if strict_coercion:
+            return self._get_dt_sc_loader(tuple_mapper)
+        return self._get_dt_non_sc_loader(tuple_mapper)
+
+    def _get_dt_non_sc_loader(self, tuple_mapper):
+        def dt_non_sc_loader(data):
+            try:
+                value_tuple = tuple(data)
+            except TypeError:
+                raise TypeLoadError(tuple, data)
+            return tuple(tuple_mapper(value_tuple))
+
+        return dt_non_sc_loader
+
+    def _get_dt_sc_loader(self, tuple_mapper):
+        def dt_sc_loader(data):
+            if isinstance(data, CollectionsMapping):
+                raise ExcludedTypeLoadError(tuple, Mapping, data)
+            if type(data) is str:  # pylint: disable=unidiomatic-typecheck
+                raise ExcludedTypeLoadError(tuple, str, data)
+            try:
+                value_tuple = tuple(data)
+            except TypeError:
+                raise TypeLoadError(tuple, data)
+            return tuple(tuple_mapper(value_tuple))
+
+        return dt_sc_loader
+
+    def _create_dt_all_loader(self, loaders: Collection[Loader]):
+        def dt_all_loader(data):
+
+            try:
+                data_len = len(data)
+            except TypeError:
+                raise TypeLoadError(tuple, data)
+
+            loaders_len = len(loaders)
```

Comment:
loaders_len could be calculated before loader start working, when closure is created